### PR TITLE
Support supplementaryViews created from nibs

### PR DESCRIPTION
**Pull request by rawlinxx**
## Changes in this pull request

Summary: add dequeueReusableSupplementaryView with nib

### CODE REVIEW

**Comment by rnystrom** [comment]
@rawlinxx this is awesome! Thanks for taking this on.

In order to merge this, we're going to have to add some unit tests. Take a look at #56 for an example of requirements for this.

File: Source/IGListCollectionContext.h [hunk 1]
```diff
@@ -126,6 +126,25 @@ NS_ASSUME_NONNULL_BEGIN
...
+ @param index             The index of the supplementary vew.
+ 
+ @return A supplementary view dequeued from the reuse pool or newly created.
+ 
+ @note This method uses a string representation of the view class as the identifier.
+ */
+- (__kindof UICollectionReusableView *)dequeueReusableSupplementaryViewOfKind:(NSString *)elementKind
+                                                         forSectionController:(IGListSectionController <IGListSectionType> *)sectionController
+                                                                      nibName:(NSString *)nibName
+                                                                       bundle:(NSBundle *)bundle
```
Comment (rnystrom) [inline Source/IGListCollectionContext.h]: We'll need this to be nullable: `bundle:(nullable NSBundle *)bundle`

File: Source/IGListAdapter.m [hunk 2]
```diff
@@ -759,6 +760,23 @@ - (__kindof UICollectionReusableView *)dequeueReusableSupplementaryViewOfKind:(N
     return [collectionView dequeueReusableSupplementaryViewOfKind:elementKind withReuseIdentifier:identifier forIndexPath:indexPath];
 }
 
+- (__kindof UICollectionReusableView *)dequeueReusableSupplementaryViewOfKind:(NSString *)elementKind
+                                                         forSectionController:(IGListSectionController <IGListSectionType> *)sectionController
+                                                                      nibName:(NSString *)nibName
+                                                                       bundle:(NSBundle *)bundle
+                                                                      atIndex:(NSInteger)index {
+    IGAssertMainThread();
```
Comment (rnystrom) [inline Source/IGListAdapter.m]: Let's also add:
```objc
IGParameterAssert([nibName length] > 0);
IGParameterAssert([elementKind length] > 0);
```
I realize those are also missing from other methods. Adding some starter tasks for that.

File: Source/IGListStackedSectionController.m [hunk 3]
```diff
@@ -219,6 +219,20 @@ - (UICollectionReusableView *)dequeueReusableSupplementaryViewOfKind:(NSString *
...
 }
 
+- (__kindof UICollectionReusableView *)dequeueReusableSupplementaryViewOfKind:(NSString *)elementKind
+                                                         forSectionController:(IGListSectionController <IGListSectionType> *)sectionController
+                                                                      nibName:(NSString *)nibName
+                                                                       bundle:(NSBundle *)bundle
+                                                                      atIndex:(NSInteger)index
+{
+    const NSUInteger offset = [self offsetForSectionController:sectionController];
+    return (UICollectionViewCell *_Nonnull)[self.collectionContext dequeueReusableSupplementaryViewOfKind:elementKind
```
Comment (rnystrom) [inline Source/IGListStackedSectionController.m]: Do we need this cast?

File: Source/IGListAdapter.m (hunk 2 above)
Comment (jessesquires) [inline Source/IGListAdapter.m]: `bundle` should be `nullable` here, right?

**Comment by jessesquires** [comment]
Also need to update the `CHANGELOG.md` -- add an entry under the `2.0.0` release 😄

File: Source/IGListAdapter.m (hunk 2 above)
Comment (rawlinxx) [inline Source/IGListAdapter.m]: Sorry for that, I'm trying this framework, hope to have this feature ASAP, without attention to details

File: Source/IGListCollectionContext.h (hunk 1 above)
Comment (rawlinxx) [inline Source/IGListCollectionContext.h]: right! just like the documentation i copied said :)

File: Source/IGListStackedSectionController.m (hunk 3 above)
Comment (rawlinxx) [inline Source/IGListStackedSectionController.m]: I do not quite understand why (UICollectionViewCell *_Nonnull) is needed here, just following the previous method implementation since I have not read all the source code yet, i think UICollectionReusableView is more suitable here, am i wrong?

File: Source/IGListAdapter.m (hunk 2 above)
Comment (rnystrom) [inline Source/IGListAdapter.m]: @rawlinxx no worries! PRs are an iterative process, its good to get something up and we can all work on it together 👨‍👩‍👧‍👦

File: Source/IGListAdapter.m [hunk 4]
```diff
@@ -759,6 +761,25 @@ - (__kindof UICollectionReusableView *)dequeueReusableSupplementaryViewOfKind:(N
     return [collectionView dequeueReusableSupplementaryViewOfKind:elementKind withReuseIdentifier:identifier forIndexPath:indexPath];
 }
 
+- (__kindof UICollectionReusableView *)dequeueReusableSupplementaryViewOfKind:(NSString *)elementKind
+                                                         forSectionController:(IGListSectionController <IGListSectionType> *)sectionController
```
Comment (rnystrom) [inline Source/IGListAdapter.m]: nit: let's remove the space between `IGListSectionController<IGListSectionType>`

File: Source/IGListCollectionContext.h [hunk 5]
```diff
@@ -126,6 +126,25 @@ NS_ASSUME_NONNULL_BEGIN
...
+ @param nibName           The name of the nib file.
+ @param bundle            The bundle in which to search for the nib file. If nil, this method looks for the nib file in the main bundle.
+ @param index             The index of the supplementary vew.
+ 
+ @return A supplementary view dequeued from the reuse pool or newly created.
+ 
+ @note This method uses a string representation of the view class as the identifier.
+ */
+- (__kindof UICollectionReusableView *)dequeueReusableSupplementaryViewOfKind:(NSString *)elementKind
+                                                         forSectionController:(IGListSectionController <IGListSectionType> *)sectionController
```
Comment (rnystrom) [inline Source/IGListCollectionContext.h]: nit: spacing between class and protocol

File: Source/IGListStackedSectionController.m [hunk 6]
```diff
@@ -213,12 +213,26 @@ - (UICollectionReusableView *)dequeueReusableSupplementaryViewOfKind:(NSString *
...
     const NSUInteger offset = [self offsetForSectionController:sectionController];
-    return (UICollectionViewCell *_Nonnull)[self.collectionContext dequeueReusableSupplementaryViewOfKind:elementKind
+    return (UICollectionReusableView *_Nonnull)[self.collectionContext dequeueReusableSupplementaryViewOfKind:elementKind
                                                                                      forSectionController:self
                                                                                                     class:viewClass
                                                                                                   atIndex:(index + offset)];
 }
 
+- (UICollectionReusableView *)dequeueReusableSupplementaryViewOfKind:(NSString *)elementKind
+                                                forSectionController:(IGListSectionController <IGListSectionType> *)sectionController
```
Comment (rnystrom) [inline Source/IGListStackedSectionController.m]: spacing between class/protocol

File: Source/IGListStackedSectionController.m [hunk 7]
```diff
@@ -213,12 +213,26 @@ - (UICollectionReusableView *)dequeueReusableSupplementaryViewOfKind:(NSString *
...
+                                                             nibName:(NSString *)nibName
+                                                              bundle:(NSBundle *)bundle
+                                                             atIndex:(NSInteger)index
+{
+    const NSUInteger offset = [self offsetForSectionController:sectionController];
+    return (UICollectionReusableView *_Nonnull)[self.collectionContext dequeueReusableSupplementaryViewOfKind:elementKind
+                                                                                     forSectionController:self
+                                                                                                  nibName:nibName
+                                                                                                   bundle:bundle
+                                                                                                  atIndex:(index + offset)];
```
Comment (rnystrom) [inline Source/IGListStackedSectionController.m]: can we realign the colons?

File: Source/IGListStackedSectionController.m [hunk 8]
```diff
@@ -213,12 +213,26 @@ - (UICollectionReusableView *)dequeueReusableSupplementaryViewOfKind:(NSString *
...
                                                                                                     class:viewClass
                                                                                                   atIndex:(index + offset)];
 }
 
+- (UICollectionReusableView *)dequeueReusableSupplementaryViewOfKind:(NSString *)elementKind
+                                                forSectionController:(IGListSectionController <IGListSectionType> *)sectionController
+                                                             nibName:(NSString *)nibName
+                                                              bundle:(NSBundle *)bundle
+                                                             atIndex:(NSInteger)index
+{
```
Comment (rnystrom) [inline Source/IGListStackedSectionController.m]: curly needs to be on the same line as the last param

File: Source/IGListStackedSectionController.m [hunk 9]
```diff
@@ -213,12 +213,26 @@ - (UICollectionReusableView *)dequeueReusableSupplementaryViewOfKind:(NSString *
...
+    const NSUInteger offset = [self offsetForSectionController:sectionController];
+    return (UICollectionReusableView *_Nonnull)[self.collectionContext dequeueReusableSupplementaryViewOfKind:elementKind
```
Comment (rnystrom) [inline Source/IGListStackedSectionController.m]: > i think UICollectionReusableView is more suitable here, am i wrong?

Hmm I'm actually not sure why we're casting it in the other methods. I just tried it locally and we can remove those. I wouldn't worry about it now. Can we revert the `UICollectionReusableView` changes? We can do cleanup in another PR.

Comment (rawlinxx) [inline Source/IGListStackedSectionController.m]: done

**Comment by rnystrom** [comment]
@rawlinxx Circling back to this, we wont be able to land until we add some unit tests. #56 added nib support for cells along w/ some unit tests. If we can basically copy exactly what we added there we'll be able to merge!

**Comment by zhubofei** [comment]
Any update on this PR 🤔?

**Comment by jessesquires** [comment]
Re-opening. @rawlinxx - was this closed by mistake?

**Comment by jessesquires** [comment]
@rnystrom - If needed, we can push commits directly to this PR for tests and carry it forward.

**Comment by jessesquires** [comment]
We're carrying these changes forward with some small modifications/additions in #162. 

Closing in favor of #162.
